File: process/Python/data/input.py
from pyarrow.parquet import read_table as pq_read_table
from pyarrow.parquet import write_table as pq_write_table
import pyarrow as pa
from os.path import join
from pandas import DataFrame
from copy import deepcopy
from itertools import product as iter_product
from process.Python.model.xgboost import xgboost_model
from logging import getLogger
from process.Python.model.tax_calculator import cal_tax

logger = getLogger()


def add_initial_pop_status(pop: DataFrame) -> DataFrame:
    pop["life_stage"] = "alive"

    return pop


def create_inputs(
    data_dir: str,
    required_data_types: list = ["pop"],
    data_type: str = "parquet",
    base_year: int or None = None,
):
    inputs = {}
    for proc_data_type in required_data_types:
        proc_data_path = join(data_dir, f"{proc_data_type}_data.{data_type}")
        proc_table = pq_read_table(proc_data_path)
        proc_table = proc_table.to_pandas()

        if proc_data_type == "pop":
            proc_table = add_initial_pop_status(proc_table)

        if base_year is not None:
            proc_table["base_year"] = base_year

        inputs[proc_data_type] = proc_table

    return inputs


def prepare_ruf_inputs(
    df_input: DataFrame,
    hours_options: list,
    total_hours: int,
    leisure_value: float,
    market_income_name_per_hour: str,
    benefit_income_name_per_week: str,
    working_hours_name: str,
    data_scaler: float = 1000.0,
    data_output_path: str = None,
):

    # select relevant features
    df_processed = deepcopy(
        df_input[
            [
                "id",
                "household_id",
                "selected",
                market_income_name_per_hour,
                benefit_income_name_per_week,
                working_hours_name,
            ]
        ]
    )

    all_household_ids = df_processed["household_id"].unique()
    long_data = []
    for index1, proc_hhld_id in enumerate(all_household_ids):

        if index1 % 10 == 0:
            logger.info(
                f"Processing input: {round(100 * index1 / len(all_household_ids), 3)}%"
            )

        proc_hhld = df_processed[df_processed["household_id"] == proc_hhld_id]

        num_people = len(proc_hhld)
        all_possible_hours_combination = list(
            iter_product(hours_options, repeat=num_people)
        )

        chosen_combination = min(
            all_possible_hours_combination,
            key=lambda c: sum(
                (a - b) ** 2 for a, b in zip(c, list(proc_hhld["working_hours"].values))
            ),
        )

        for i_comb, proc_combination in enumerate(all_possible_hours_combination):

            is_chosen = 0
            if proc_combination == chosen_combination:
                is_chosen = 1

            proc_data_list = []
            proc_hhld_market_income = 0
            proc_hhld_benefit_income = 0
            for index2, proc_hours in enumerate(proc_combination):
                proc_person = proc_hhld.iloc[index2]
                person_wage = proc_person[market_income_name_per_hour]
                leisure_hours = total_hours - proc_hours

                market_income_person = person_wage * proc_hours
                benefit_income_person = proc_person[benefit_income_name_per_week]

                proc_hhld_market_income += market_income_person
                proc_hhld_benefit_income += benefit_income_person

                if proc_person["selected"]:
                    proc_data_list.append(
                        {
                            "option_hours_id": i_comb,
                            "household_id": int(proc_hhld["household_id"].values[0]),
                            "people_id": int(proc_person["id"]),
                            "option_hours": proc_hours,
                            "is_chosen": is_chosen,
                            "market_income": market_income_person,
                            "benefit_income": benefit_income_person,
                            "leisure": leisure_hours * leisure_value,
                        }
                    )

            for proc_data in proc_data_list:
                proc_data["market_income_hhld"] = proc_hhld_market_income
                proc_data["benefit_income_hhld"] = proc_hhld_benefit_income
            long_data.extend(proc_data_list)

    results = DataFrame(long_data)
    income_cols = [
        "market_income",
        "benefit_income",
        "market_income_hhld",
        "benefit_income_hhld",
    ]
    results[income_cols] = results[income_cols].clip(lower=1e-9)
    # results = results.drop_duplicates()
    results = obtain_benefit_income(results)

    # make the data scaled: e.g., income is between 0-2, leisure is between 0-1
    for proc_key in ["income", "income_hhld", "leisure"]:
        results[proc_key] = results[proc_key] / data_scaler
    results["leisure"] = results["leisure"] * 23.0

    results = results.sort_values(
        by=["household_id", "people_id", "option_hours"]
    ).reset_index(drop=True)

    pq_write_table(pa.Table.from_pandas(results), data_output_path)


def obtain_benefit_income(df: DataFrame) -> DataFrame:

    logger.info("Predict benefit income using XGBoost model")

    df_for_train = df[df["is_chosen"] == 1].reset_index(drop=True)
    model = xgboost_model(
        df_for_train, "benefit_income", ["market_income", "market_income_hhld"]
    )
    import xgboost as xgb

    dpredict = xgb.DMatrix(
        df[["market_income", "market_income_hhld"]].values.astype("float32")
    )
    df["benefit_income_pred"] = model.predict(dpredict)
    # df["benefit_income_pred"] = model.predict(
    #    df[["market_income", "market_income_hhld"]]
    # )

    df["income"] = df["market_income"] + df["benefit_income_pred"]
    df["tax"] = cal_tax(df["income"])
    df["income"] = df["income"] - df["tax"]
    df["income_hhld"] = df.groupby(["option_hours_id", "household_id"])[
        "income"
    ].transform("sum")
    df = df[
        [
            "option_hours_id",
            "household_id",
            "people_id",
            "option_hours",
            "is_chosen",
            "leisure",
            "market_income",
            "benefit_income_pred",
            "income",
            "income_hhld",
        ]
    ]
    return df


"""
import matplotlib.pyplot as plt

plt.scatter(
    df["market_income_hhld"],
    df["benefit_income_pred"],
    alpha=0.3,
    s=5,
)
plt.scatter(
    df["market_income_hhld"],
    df["benefit_income"],
    alpha=0.3,
    c="r",
    s=5,
)
plt.savefig("test.png")
plt.close()
df2 = df[
    [
        "people_id",
        "household_id",
        "benefit_income",
        "benefit_income_pred",
        "market_income",
        "market_income_hhld",
    ]
]
"""

File: process/R/data/input.R


add_initial_pop_status <- function(pop) {
  pop[, life_stage := "alive"]
  return(pop)
}

create_inputs <- function(data_dir, required_data_types = c("pop"), data_type = "parquet", base_year = NULL) {
  
  inputs <- list()
  for (proc_data_type in required_data_types) {
    proc_data_path <- paste0(
      data_dir, "/", paste0(proc_data_type, "_data.", data_type)
    )
    proc_data <- read_parquet(proc_data_path)
    
    proc_data <- as.data.table(proc_data)
    
    if (proc_data_type == "pop") {
      proc_data = add_initial_pop_status(proc_data)
    }
    
    if (! is.null(base_year)) {
      proc_data$base_year <- base_year
    }
    
    inputs[[proc_data_type]] <- as.data.frame(proc_data)
  }
  
  return (inputs)
}

prepare_ruf_inputs <- function(df_input, 
                               hours_options, 
                               total_hours, 
                               leisure_value, 
                               market_income_name_per_hour, 
                               benefit_income_name_per_week, 
                               working_hours_name, 
                               data_scaler = 1000.0, 
                               data_output_path = NULL) {
  
  # 1. Setup data.table and select relevant columns
  # Using copy to avoid side effects on the input dataframe
  dt <- as.data.table(df_input)[, .(
    id, 
    household_id, 
    selected, 
    market_hour = get(market_income_name_per_hour),
    benefit_week = get(benefit_income_name_per_week),
    working_hours = get(working_hours_name)
  )]
  
  # 2. Process by Household
  # We use a split-apply-combine approach which is highly optimized in data.table
  results <- dt[, {
    num_people <- .N
    # Generate all possible hours combinations for the household size
    # CJ (Cross Join) is the data.table equivalent of itertools.product
    hours_list <- replicate(num_people, hours_options, simplify = FALSE)
    combs <- do.call(CJ, hours_list)
    setnames(combs, paste0("V", 1:num_people))
    
    # Add an option ID
    combs[, option_hours_id := .I - 1]
    
    # 3. Find 'chosen_combination' (Nearest Neighbor via Euclidean Distance)
    # Vectorized distance calculation across rows of combinations
    obs_hours <- .SD$working_hours
    # Calculate sum of squared differences for each row
    dists <- rowSums(mapply(function(col, obs) (col - obs)^2, combs[, 1:num_people, with=FALSE], obs_hours))
    chosen_idx <- which.min(dists)
    combs[, is_chosen := as.integer(.I == chosen_idx)]
    
    # 4. Calculate Household Totals for each combination
    # Calculate market income for each person (matrix multiplication for speed)
    wages <- .SD$market_hour
    person_benefits <- .SD$benefit_week
    
    # Calculate hhld totals for each combination (rowSums)
    market_incomes_matrix <- mapply(`*`, combs[, 1:num_people, with=FALSE], wages)
    combs[, market_income_hhld := rowSums(as.matrix(market_incomes_matrix))]
    combs[, benefit_income_hhld := sum(person_benefits)] # Based on Python logic
    
    # 5. Transform to Long Format (People within Options)
    # We only keep rows where 'selected' is TRUE (matching proc_person["selected"])
    selected_indices <- which(.SD$selected)
    
    # Melt the combinations to person-level
    long_combs <- melt(combs, 
                       id.vars = c("option_hours_id", "is_chosen", "market_income_hhld", "benefit_income_hhld"),
                       measure.vars = paste0("V", 1:num_people),
                       variable.name = "person_idx", 
                       value.name = "option_hours")
    
    # Map person_idx back to actual data
    long_combs[, person_idx := as.numeric(gsub("V", "", person_idx))]
    
    # Filter for 'selected' people only
    long_combs <- long_combs[person_idx %in% selected_indices]
    
    # Join with original person attributes
    person_map <- .SD[, .(person_idx = .I, people_id = id)]
    long_combs <- merge(long_combs, person_map, by = "person_idx")
    
    # 6. Final calculations for the group
    long_combs[, `:=`(
      market_income = wages[person_idx] * option_hours,
      benefit_income = person_benefits[person_idx],
      leisure = (total_hours - option_hours) * leisure_value
    )]
    
    long_combs[, .(option_hours_id, people_id, 
                   option_hours, is_chosen, market_income, benefit_income, 
                   leisure, market_income_hhld, benefit_income_hhld)]
    
  }, by = household_id]
  
  # 7. Post-processing (Clipping and Scaling)
  income_cols <- c("market_income", "benefit_income", "market_income_hhld", "benefit_income_hhld")
  for (col in income_cols) {
    set(results, i = which(results[[col]] < 1e-9), j = col, value = 1e-9)
  }
  # Call external benefit function (ensure this is defined in your environment)
  results <- obtain_benefit_income(results)
  setDT(results)
  
  # Scaling
  # Note: Python code defines 'income' and 'income_hhld' which weren't in the list above.
  # Assuming they are created by obtain_benefit_income.
  scale_cols <- intersect(c("income", "income_hhld", "leisure"), names(results))
  for (col in scale_cols) {
    results[, (col) := get(col) / data_scaler]
  }
  
  # Apply the specific leisure multiplier from your Python code
  results[, leisure := leisure * 23.0]
  
  # 8. Sort and Save
  setorder(results, household_id, people_id, option_hours)
  
  if (!is.null(data_output_path)) {
    write_parquet(results, data_output_path)
  }
  
  return(results)
}


obtain_benefit_income <- function(df) {
  print("Predict benefit income using XGBoost model")
  
  df_for_train <- df %>% filter(is_chosen == 1)
  
  # Prepare data for XGBoost
  train_features <- as.matrix(df_for_train[, c("market_income", "market_income_hhld")])
  train_label <- df_for_train$benefit_income
  
  model <- model_xgboost_env$xgboost_model(df_for_train,
                                  "benefit_income",
                                  c("market_income", "market_income_hhld"))
  # Prediction
  full_features <- as.matrix(df[, c("market_income", "market_income_hhld")])
  df$benefit_income_pred <- predict(model, full_features)
  
  # Calculate after-tax income
  df <- df %>%
    mutate(
      income = market_income + benefit_income_pred,
      tax = model_tax_calculator_env$cal_tax(income), # Assuming cal_tax is defined globally
      income = income - tax
    ) %>%
    group_by(option_hours_id, household_id) %>%
    mutate(income_hhld = sum(income)) %>%
    ungroup() %>%
    select(
      option_hours_id,
      household_id,
      people_id,
      option_hours,
      is_chosen,
      leisure,
      market_income,
      benefit_income_pred,
      income,
      income_hhld
    )
  
  return(df)
}
